### ISSUE TITLE
code.ipynb

### ISSUE THREAD

**Issue by kauecodify** [kauecodify/socioeconomic-analysis-of-the-state-of-minas#1]
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import warnings

warnings.filterwarnings("ignore", category=FutureWarning)

dados_tratados = pd.read_csv('/content/drive/MyDrive/*way*/dados_tratados.csv', sep=',', encoding='iso-8859-1')

from google.colab import drive
drive.mount('/content/drive')

Transformar variáveis qualitativas em quantitativas (Optativo)

#dados_tratados.head
# dados_tratados.Q025.replace(['A', 'B'], [0, 1], inplace=True)
# dados_tratados.head()
# dados_tratados.dtypes
# dados_tratados.Q024.replace(['A', 'B','C','D','E'], [0, 1, 2, 3, 4], inplace=True)
# dados_tratados.head()
# dados_tratados.dtypes
# dados_tratados.shape

Preparação dos atributos - Separar os atributos previsores do atributo classe (resultado)

dados_tratados.head()
previsores = dados_tratados.iloc[:, 0:17].values
classe = dados_tratados.iloc[:, 17].values

Transformar variáveis categóricas em variáveis discretas (LabelEncoder)
OneHotEncoder: variáveis dummy

from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.compose import ColumnTransformer
ColumnTransformer?

Parâmetros ColumnTransformer

#name: nome dado a transformação.
#transformer: tipo de estimador (OneHotEncoder).
#columns: colunas que serão transformadas.
#remainder: o que acontecerá com o restante das colunas não relacionadas: 1) drop = exclui as outras colunas. 2) passthrough = mantém as outras colunas. drop é default.
#sparse_threshold: parâmetro de classificação de matrizes esparsas. default é 0.3
#n_jobs: número de trabalhos a serem executados em paralelo. default é nenhum
#transformer_weights: definição de pesos aos transformadores.
#verbose: default é False. se for True a execução é apresentada na tela.

transformacao_previsores = ColumnTransformer([('transformacao', OneHotEncoder(), [0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16])],remainder='passthrough')
previsores = transformacao_previsores.fit_transform(previsores).toarray()
previsores
previsores.shape
classe
transformacao_classe = LabelEncoder()
classe = transformacao_classe.fit_transform(classe)
classe
dados_tratados.head()

Escalonamento: alterando escala (se necessário).
Útil quando os valores são muito diferentes. A interpretação fica comprometida.

# from sklearn.preprocessing import StandardScaler
# scaler = StandardScaler()
# previsores = scaler.fit_transform(previsores)
# previsores

Algoritmo com base de treinamento e teste no Scikit-Learn

from sklearn.model_selection import train_test_split

Parâmetros train_test_split:

#arrays: nomes dos atributos previsores e alvo.
#test_size: tamanho em porcentagem dos dados de teste. default é none.
#train_size: tamanho em porcentagem dos dados de treinamento.default é none.
#random_state: nomeação de um estado aleatório.
#shuffle: embaralhamento dos dados aleatórios. Associado com o random_state #ocorre o mesmo embaralhamento sempre. Default é True.
#stratify: Possibilidade de dividir os dados de forma estratificada. Default é None (#nesse caso é mantido a proporção, isto é, se tem 30% de zeros e 70% de 1 no #dataframe, na separação em treinamento e teste se manterá essa proporção.

previsores_treinamento, previsores_teste, classe_treinamento, classe_teste = train_test_split(previsores, classe, test_size=0.15, random_state=0)

previsores_treinamento.shape
previsores_teste.shape
classe_treinamento.shape
classe_teste.shape
from sklearn.neural_network import MLPClassifier
MLPClassifier

Parâmetros MLPClassifier

hidden_layer_sizes (camadas escondidas): default (100,)

Quant.= (Ne+Ns)/2 = (105+1)/2 = 53 neurônios
Quant.=2/3.(Ne) + Ns = 2/3.105+1 = 71 neurônios

activation: Função de ativação default='relu'

solver: algoritmo matemático. Default='adam' (datasets grandes = acima de 1000 amostras). lbfgs é para datasets pequenos. sgd é com a descida do gradiente estocástico (recomendado testar).

alpha: parâmetro para o termo de regularização de ajuste de pesos. Aumento de alpha estimula pesos menores e diminuição de alpha estimula pesos maiores. Default=0.0001.

batch_size: tamanho dos mini lotes. default=min(200, n_samples). Não usar com o solver lbfgs.

learning_rate: taxa de aprendizagem. default='constant'. Três tipos:
1- 'constant':uma taxa de aprendizado constante dada pela taxa de aprendizagem inicial.
2- 'invscaling': diminui gradualmente por: taxa efetiva = taxa inicial / t^power_t
3- 'adaptive': a taxa é dividida por 5 cada vez que em duas épocas consecutivas não diminuir o erro.

learning_rate_init: taxa de aprendizagem inicial. Default=0.001

power_t: default=0.5. usado apenas para solver 'sgd'.

max_iter int: Número máximo de iterações. default = 200. ('sgd', 'adam').

max_fun: Número máximo de chamadas de função de perda. Para 'lbfgs'. Default: 15000

shuffle: default = True Usado apenas quando solver = 'sgd' ou 'adam'.

random_state: default = None

tol:Tolerância para a otimização.Default=0.0001

momentum: otimização do algoritmo 'sgd'. Default: 0.9.

n_iter_no_change: Número máximo de épocas que não atinge a tolerância de melhoria. default = 10. Apenas para solver = 'sgd' ou 'adam'

verbose : Mostra o progresso. default=False.

classificador = MLPClassifier(hidden_layer_sizes=(53,), activation='relu', solver='adam', max_iter=500,
                              tol=0.0001, random_state = 12, verbose = True)

classificador.fit(previsores_treinamento, classe_treinamento)

previsoes = classificador.predict(previsores_teste)

previsoes

print(classificador.out_activation_)

from sklearn.metrics import confusion_matrix, accuracy_score

acuracia = accuracy_score(classe_teste, previsoes)

acuracia

matriz = confusion_matrix(classe_teste, previsoes)

matriz

Algoritmo com validação cruzada no Scikit-Learn

from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.neural_network import MLPClassifier

Parâmetros StratifiedkFold:

n_splits (k): número de divisões (folds). recomendado=10. Default = 5.

kfold = StratifiedKFold(n_splits = 10, shuffle = True, random_state = 3)

exemplo = np.zeros(shape=(previsores.shape[0], 1)) 

#exemplo para criar índices para fazer a separação treinamento e teste

exemplo

exemplo.shape

resultados = []
matrizes = []
for indice_treinamento, indice_teste in kfold.split(previsores,
                                                    np.zeros(shape=(previsores.shape[0], 1))): # cria índices para todos os registros (linhas) para fazer a separação;
    classificador = MLPClassifier(hidden_layer_sizes=(53,), max_iter=500, tol=0.001, random_state = 12, verbose = True)
    classificador.fit(previsores[indice_treinamento], classe[indice_treinamento])
    previsoes = classificador.predict(previsores[indice_teste])
    acuracia = accuracy_score(classe[indice_teste], previsoes)
    matrizes.append(confusion_matrix(classe[indice_teste], previsoes))
    resultados.append(acuracia)

resultados = np.asarray(resultados)
resultados.mean() #média

resultados.std() #desvio padrão

matriz_final = np.mean(matrizes, axis = 0)

matriz_final

Algoritmo com base de teste e treinamento no Keras

import keras
from keras.models import Sequential
from keras.layers import Dense
from sklearn.model_selection import train_test_split

previsores_treinamento, previsores_teste,classe_treinamento, classe_teste = train_test_split(previsores, classe, test_size=0.15, random_state=12)

previsores.shape

previsores_treinamento.shape

classificador = Sequential()

units: neurônios de saída de cada camada.
ativação : função de ativação a ser usada. Default="linear"
use_bias: Se a camada usa um vetor de polarização. Default=True
kernel_initializer: inicializador para a kernel matriz de pesos. Default='glorot_uniform'
bias_initializer: inicializador para o vetor de polarização. Default='zeros'
kernel_regularizer: função regularizadora aplicada à matriz de pesos.
bias_regularizer: função regularizadora aplicada ao vetor de polarização.
activity_regularizer: Função regularizadora aplicada à saída da camada (sua "ativação").
kernel_constraint : função de restrição aplicada ao kernel matriz de pesos.
bias_constraint : função de restrição aplicada ao vetor de polarização.

classificador.add(Dense(units = 53, activation = 'relu', input_dim = 105))

classificador.add(Dense(units = 53, activation = 'relu'))

classificador.add(Dense(units = 1, activation = 'sigmoid'))

classificador.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

classificador.fit(previsores_treinamento, classe_treinamento, batch_size = 10, epochs = 8)

previsoes = classificador.predict(previsores_teste)

previsoes

previsoes = (previsoes > 0.5)

previsoes

from sklearn.metrics import confusion_matrix, accuracy_score

acuracia = accuracy_store(classe_teste, previsoes)

matriz
